File: p2pnet/android/app/src/main/java/com/example/p2pnet/data/repository/P2pRepository.kt
package com.example.p2pnet.data.repository

import android.util.Log
import com.example.p2pnet.data.local.LocalPrefs
import com.example.p2pnet.data.remote.WsClient
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONObject
import kotlin.coroutines.resume

class P2pRepository(
    private val localPrefs: LocalPrefs
) {
    private var _client: WsClient = WsClient(localPrefs)
    private var _loggedInUsername: String? = null
    val loggedInUsername: String? get() = _loggedInUsername

    sealed class LoginResult {
        data class Success(val username: String) : LoginResult()
        data class Error(val message: String) : LoginResult()
    }

    var onRawMessage: ((String) -> Unit)? = null
    var onSend: ((String) -> Unit)? = null
    var onUdpSend: ((String) -> Unit)? = null
    var onUdpRecv: ((String) -> Unit)? = null
    var onHelloDone: ((WsClient.PeerInfo?, java.net.DatagramSocket?, String, Int, String) -> Unit)? = null

    fun useClient(client: WsClient?) {
        _client = client ?: WsClient(localPrefs)
    }

    fun getClient(): WsClient = _client

    private val ws: WsClient get() = _client

    suspend fun login(useWss: Boolean, host: String, port: Int, username: String, password: String): LoginResult {
        _loggedInUsername = username
        localPrefs.serverHost = host
        localPrefs.serverPort = port
        localPrefs.username = username
        localPrefs.loggedIn = true

        return suspendCancellableCoroutine { cont ->
            val listener = object : WsClient.Listener {
                override fun onConnected() {}

                override fun onSend(text: String) {
                    this@P2pRepository.onSend?.invoke(text)
                }

                override fun onRawMessage(text: String) {
                    this@P2pRepository.onRawMessage?.invoke(text)
                }

                override fun onUdpSend(text: String) {
                    this@P2pRepository.onUdpSend?.invoke(text)
                }

                override fun onUdpRecv(text: String) {
                    this@P2pRepository.onUdpRecv?.invoke(text)
                }

                override fun onHelloDone(info: WsClient.PeerInfo?, sock: java.net.DatagramSocket?, peerIp: String, peerPort: Int, mode: String) {
                    this@P2pRepository.onHelloDone?.invoke(info, sock, peerIp, peerPort, mode)
                }

                override fun onLoginSuccess(username: String) {
                    _loggedInUsername = username
                    localPrefs.loggedIn = true
                    cont.resume(LoginResult.Success(username))
                }

                override fun onLoginFailed(message: String) {
                    localPrefs.clearSession()
                    _loggedInUsername = null
                    cont.resume(LoginResult.Error(message))
                }

                override fun onDisconnected() {
                    if (cont.isActive) {
                        cont.resume(LoginResult.Error("连接断开"))
                    }
                }

                override fun onError(message: String) {
                    if (cont.isActive) {
                        cont.resume(LoginResult.Error(message))
                    }
                }
            }

            ws.listener = listener
            ws.login(useWss, host, port, username, password)

            cont.invokeOnCancellation {
                ws.disconnectOnly()
            }
        }
    }

    fun logout() {
        ws.disconnectOnly()
        localPrefs.clearSession()
        _loggedInUsername = null
    }

    fun connectOnly(useWss: Boolean, host: String, port: Int) {
        val proto = if (useWss) "wss" else "ws"
        ws.connect("$proto://$host:$port/")
    }

    fun disconnectOnly() {
        ws.disconnectOnly()
    }

    fun resetUdpState() {
        ws.resetUdpState()
    }

    fun sendList() = ws.sendList()
    fun sendWghelp(target: String) = ws.sendWghelp(target)
    fun sendP2pUdp(target: String) = ws.sendP2pUdp(target)
    fun sendP2pTcp(target: String) = ws.sendP2pTcp(target)
    fun getServerHost(): String = localPrefs.serverHost
    fun getServerPort(): Int = localPrefs.serverPort
    fun isLoggedIn(): Boolean = localPrefs.loggedIn
    fun getSavedUsername(): String? = localPrefs.username
}

File: p2pnet/ios/p2pnet/Data/Remote/P2pRepository.swift
import Foundation

enum LoginResult {
    case success(String)
    case error(String)
}

class P2pRepository {
    private var _client: WsClient
    private var _connectionListener: WsClientListenerWrapper?
    private var _loggedInUsername: String?
    var loggedInUsername: String? { _loggedInUsername }

    // MARK: - Stored callbacks (renamed to avoid conflict with protocol methods)
    var onLogMessage: ((String) -> Void)?
    var onConnectedHandler: (() -> Void)?
    var onDisconnectedHandler: (() -> Void)?
    var onError: ((String) -> Void)?

    var onSend: ((String) -> Void)?
    var onUdpSend: ((String) -> Void)?
    var onUdpRecv: ((String) -> Void)?
    var onRecv: ((String) -> Void)?
    var onHelloDone: ((PeerInfo?, Int32?, String, Int, String) -> Void)?
    // Login-specific callbacks (set by login())
    private var loginSuccessHandler: ((String) -> Void)?
    private var loginFailedHandler: ((String) -> Void)?
    private var loginDisconnectedHandler: (() -> Void)?



    private let localPrefs: LocalPrefs

    init(localPrefs: LocalPrefs) {
        self.localPrefs = localPrefs
        _client = WsClient()
    }

    func useClient(_ client: WsClient?) {
        _client = client ?? WsClient()
    }

    func getClient() -> WsClient {
        return _client
    }

    func setupConnectionCallbacks() {
        let wrapper = WsClientListenerWrapper(
            onMessage: { [weak self] in self?.onLogMessage?($0) },
            onRecv: { [weak self] in self?.onRecv?($0) },
            onSend: { [weak self] in self?.onSend?($0) },
            onUdpSend: { [weak self] in self?.onUdpSend?($0) },
            onUdpRecv: { [weak self] in self?.onUdpRecv?($0) },
            onHelloDone: { [weak self] in self?.onHelloDone?($0, $1, $2, $3, $4) },
            onLoginSuccess: { [weak self] in self?.loginSuccessHandler?($0) },
            onLoginFailed: { [weak self] in self?.loginFailedHandler?($0) },
            onError: { _ in },
            onDisconnected: { [weak self] in self?.onDisconnectedHandler?() },
            onConnected: { [weak self] in self?.onConnectedHandler?() }
        )
        print("[P2pRepo] setupConnectionCallbacks: wrapper created, stored callbacks registered")
        _connectionListener = wrapper
        print("[P2pRepo] setupConnectionCallbacks: wrapper=\(String(format: "%p", unsafeBitCast(wrapper, to: Int.self))) _client.listener will be set")
        _client.listener = wrapper
        print("[P2pRepo] setupConnectionCallbacks: done")
    }

    func login(useWss: Bool, host: String, port: Int, username: String, password: String) async -> LoginResult {
        _loggedInUsername = username
        localPrefs.serverHost = host
        localPrefs.serverPort = port
        localPrefs.username = username
        localPrefs.loggedIn = true

        return await withCheckedContinuation { cont in
            self.loginSuccessHandler = { username in
                self._loggedInUsername = username
                self.localPrefs.loggedIn = true
                cont.resume(returning: .success(username))
            }
            self.loginFailedHandler = { message in
                self.localPrefs.clearSession()
                self._loggedInUsername = nil
                cont.resume(returning: .error(message))
            }
            // Do NOT replace _client.listener - keep the connection callbacks
            _client.login(useWss: useWss, serverHost: host, serverPort: port, username: username, password: password)
        }
    }

    func logout() {
        _client.disconnectOnly()
        localPrefs.clearSession()
        _loggedInUsername = nil
    }

    func connectOnly(useWss: Bool, host: String, port: Int) {
        _client.connect(useWss: useWss, host: host, port: port)
    }

    func disconnectOnly() {
        _client.disconnectOnly()
    }

    func resetUdpState() {
        _client.resetUdpState()
    }

    func sendList() { _client.sendList() }
    func sendWghelp(_ target: String) { _client.sendWghelp(target) }
    func sendP2pUdp(_ target: String) { _client.sendP2pUdp(target) }
    func sendP2pTcp(_ target: String) { _client.sendP2pTcp(target) }
    func getServerHost() -> String { localPrefs.serverHost }
    func getServerPort() -> Int { localPrefs.serverPort }
    func isLoggedIn() -> Bool { localPrefs.loggedIn }
    func getSavedUsername() -> String? { localPrefs.username }
}

// MARK: - WsClientListenerWrapper

class WsClientListenerWrapper: WsClientListener {
    let onMessage: (String) -> Void
    let onRecv: (String) -> Void
    let onSend: (String) -> Void
    let onUdpSend: (String) -> Void
    let onUdpRecv: (String) -> Void
    let onHelloDone: (PeerInfo?, Int32?, String, Int, String) -> Void
    let onLoginSuccess: (String) -> Void
    let onLoginFailed: (String) -> Void
    let onError: (String) -> Void
    let onDisconnectedHandler: () -> Void
    let onConnectedHandler: (() -> Void)?   // 新增

    init(
        onMessage: @escaping (String) -> Void,
        onRecv: @escaping (String) -> Void,
        onSend: @escaping (String) -> Void,
        onUdpSend: @escaping (String) -> Void,
        onUdpRecv: @escaping (String) -> Void,
        onHelloDone: @escaping (PeerInfo?, Int32?, String, Int, String) -> Void,
        onLoginSuccess: @escaping (String) -> Void,
        onLoginFailed: @escaping (String) -> Void,
        onError: @escaping (String) -> Void,
        onDisconnected: @escaping () -> Void,
        onConnected: (() -> Void)? = nil    // 新增参数，默认为 nil
    ) {
        self.onMessage = onMessage
        self.onRecv = onRecv
        self.onSend = onSend
        self.onUdpSend = onUdpSend
        self.onUdpRecv = onUdpRecv
        self.onHelloDone = onHelloDone
        self.onLoginSuccess = onLoginSuccess
        self.onLoginFailed = onLoginFailed
        self.onError = onError
        self.onDisconnectedHandler = onDisconnected
        self.onConnectedHandler = onConnected  // 保存回调
    }
    func onMessage(_ text: String) { onMessage(text) }
    func onConnected() { onConnectedHandler?() }
    func onDisconnected() { onDisconnectedHandler() }
    func onRecv(_ text: String) { onRecv(text) }
    func onSend(_ text: String) { onSend(text) }
    func onLoginSuccess(_ username: String) { onLoginSuccess(username) }
    func onLoginFailed(_ message: String) { onLoginFailed(message) }
    func onError(_ message: String) { onError(message) }
    func onUdpSend(_ text: String) { onUdpSend(text) }
    func onUdpRecv(_ text: String) { onUdpRecv(text) }
    func onHelloDone(_ info: PeerInfo?, _ sock: Int32?, _ peerIp: String, _ peerPort: Int, _ mode: String) { onHelloDone(info, sock, peerIp, peerPort, mode) }
}
